### PR TITLE
CASMCMS-7676 Add documentation for IMS recipe templating

### PR DESCRIPTION
**Pull request by ecozzi-hpe**
## Summary and Scope

_Summarize what has changed. Explain why this PR is necessary. What is impacted? Is this a new feature, critical bug fix, etc?_

Add docs for creating IMS recipes that get templated during IMS create jobs. 

_Is this change backwards incompatible, backwards compatible, or a backwards compatible bugfix?_
Yes

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [CASMCMS-7676](https://jira-pro.its.hpecorp.net:8443/browse/CASMCMS-7676)

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable
- [ ] Do not merge until CSM 1.2.5 content can be merged.

### CODE REVIEW

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 1]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+
+### Build an Image from an IMS Templated Recipe
+
+The procedure to build an image from an IMS recipe that uses templating does not change. Follow the normal IMS create
+procedure, specifying the recipe's ID value in the job's `--artifact-id` parameter. The IMS job will start as normal,
+but after downloading the recipe from S3, there will be an indication that IMS is templating the recipe in the job's
+`fetch-recipe` container.
+
+```
+ncn# i logs cray-ims-812e9cda-62a8-4d1b-83a3-5891340671ff-create-78z6p -c fetch-recipe
```
Comment (rkleinman-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:303]: Use full commands in docs, not shortcuts.

**Review by rkleinman-hpe — CHANGES_REQUESTED** [review]

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 2]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+enables variables stored in the recipe to be dynamically replaced with values registered with the IMS recipe record after the 
+recipe is downloaded from S3, but before it is built by Kiwi-NG. This functionality is used by various Cray products to ensure 
+that the resulting image is built from the most correct release versions of Nexus hosted RPM repositories. 
+
+Follow the steps below to enable IMS templating within a given recipe. Note that for IMS to properly template a recipe,
+the following procedures must all be completed.
+
+### Enable Templating in IMS Recipe Archive tgz File
+
+1. Add a file named '.ims_recipe_template.yaml' to the root of the recipe archive.
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:168]:
```suggestion
1. Add a file named `.ims_recipe_template.yaml` to the root of the recipe archive. 
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 3]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+
+1. Add a file named '.ims_recipe_template.yaml' to the root of the recipe archive. 
+
+   ``` bash
+   ncn# ls -al
+   ```
+   
+   Example output:
+
+   ``` bash
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:176]:
```suggestion
   ```
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 4]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+   drwxr-xr-x 3 root root    70 Feb 22 19:05 .
+   drwxr-xr-x 3 root root    41 Feb 22 17:18 ..      
+   -rw-r--r-- 1 root root  2744 Feb  4 21:14 config.sh
+   -rw-r--r-- 1 root root 24645 Feb  4 21:15 config.xml
+   -rw-r--r-- 1 root root  1216 Feb  4 21:14 images.sh
+   -rw-r--r-- 1 root root    29 Feb  4 21:14 .ims_recipe_template.yaml
+   drwxr-xr-x 3 root root    18 Feb  4 21:15 root
+   ```
+   
+   The contents of this file must be valid yaml containing the keyword `template_files`. The `template_files` keyword
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:186]:
```suggestion
   The contents of this file must be valid YAML, containing the keyword `template_files`. The `template_files` keyword
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 5]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+   The contents of this file must be valid yaml containing the keyword `template_files`. The `template_files` keyword 
+   must contain a list of files to be templated within the image recipe archive.
+
+   ``` bash
+   ncn# cat .ims_recipe_template.yaml
+   ```
+
+   Example output:   
+
+   ```
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:195]:
```suggestion
   ```yaml
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 6]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+   template_files:
+   - config.xml
+   ```
+   
+   IMS will template the indicated files 'in place' -- that is, the listed file will be replaced with the templated copy.
+   **NOTE:** Only files listed under the template_files key will be templated by IMS.
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:201]:
```suggestion
   **NOTE:** Only files listed under the `template_files` key will be templated by IMS. 
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 7]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+   ...
+   <repository type="rpm-md" alias="csm-{{ CSM_RELEASE_VERSION }}-sle-15sp3" priority="2" imageinclude="true">
+       <source path="https://packages.local/repository/csm-{{ CSM_RELEASE_VERSION }}-sle-15sp3/"/>
+   </repository>
+   ...
+   ```
+
+   **NOTE:** The repo referenced above is for documentation purposes only and may not actually exist. 
+
+3. Create a tgz archive of the image recipe.
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:218]:
```suggestion
3. Create a `tgz` archive of the image recipe.
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 8]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+    ```bash
+    ncn# cray ims recipes create --name "My Recipe" \
+    --recipe-type kiwi-ng --linux-distribution sles15 \
+    --template-dictionary-key CSM_RELEASE_VERSION \
+    --template-dictionary-value 1.2.5
+    ```
+   
+    Example output:
+    
+    ``` bash
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:239]:
```suggestion
    ```
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 9]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+
+   ```bash
+   ncn# cray artifacts create ims recipes/$IMS_RECIPE_ID/$ARTIFACT_FILENAME $ARTIFACT_FILENAME
+   ```
+
+3. Update the IMS recipe record with the S3 path to the recipe archive.
+
+   ```bash
+   ncn# cray ims recipes update $IMS_RECIPE_ID \
+   --link-type s3 \
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:272]:
```suggestion
           --link-type s3 \
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 10]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+   --link-path s3://ims/recipes/$IMS_RECIPE_ID/$ARTIFACT_FILENAME
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:273]:
```suggestion
           --link-path s3://ims/recipes/$IMS_RECIPE_ID/$ARTIFACT_FILENAME
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 11]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+   ```
+
+   Example output:
+
+   ``` bash
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:278]:
```suggestion
   ```
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md (hunk 1 above)
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:303]:
```suggestion
ncn# kubectl -n ims logs cray-ims-812e9cda-62a8-4d1b-83a3-5891340671ff-create-78z6p -c fetch-recipe
```

File: operations/image_management/Upload_and_Register_an_Image_Recipe.md [hunk 12]
```diff
@@ -153,3 +153,171 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
...
+but after downloading the recipe from S3, there will be an indication that IMS is templating the recipe in the job's
+`fetch-recipe` container.
+
+```
+ncn# i logs cray-ims-812e9cda-62a8-4d1b-83a3-5891340671ff-create-78z6p -c fetch-recipe
+```
+
+Example Output:
+
+``` bash
```
Comment (mharding-hpe) [inline operations/image_management/Upload_and_Register_an_Image_Recipe.md:308]:
```suggestion
```
```

**Review by mharding-hpe — CHANGES_REQUESTED** [review]

**Review by jsl-hpe — APPROVED** [review]

**Comment by mharding-hpe** [comment]
Rebased PR

**Comment by mharding-hpe** [comment]
I added a couple of linting commits to resolve the failures reported by the markdown checkers.

**Review by mharding-hpe — APPROVED** [review]

**Review by rustydb — APPROVED** [review]

**Pull request merged**